batched full hash join

jbrennan333 opened the pull request:
Signed-off-by: Jim Brennan <jimb@nvidia.com>

Closes https://github.com/NVIDIA/spark-rapids/issues/3300

Initial implementation of design proposed in https://github.com/NVIDIA/spark-rapids/issues/3300#issuecomment-1015489799

This version passes all of the join integration tests, but requires further testing.  Putting this up to allow for some review.  Currently the code is still requiring a single batch for both sides, so the batched aspect has not been tested.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -511,6 +507,208 @@ class ConditionalHashJoinIterator(
+
+  // Create a boolean column to indicate which gather map rows are valid.
+  private def validIndexMask(gatherView: ColumnView): ColumnView = {
+    withResource(Scalar.fromInt(-numRightRows)) { minValue =>
+      withResource(Scalar.fromInt(numRightRows)) { maxValue =>
+        withResource(gatherView.greaterOrEqualTo(minValue)) { greaterEqualMin =>
+          withResource(gatherView.lessThan(maxValue)) { lessMax =>
+            lessMax.and(greaterEqualMin)
+          }
+        }
+      }
+    }
```
This can be simplified to a check for != Int.MinValue. cudf should not be generating any value for a gather map that is out of bounds except for Int.MinValue.

jbrennan333 replied:
Thanks!  That is much simpler.

jbrennan333 replied:
I made this change.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -511,6 +507,208 @@ class ConditionalHashJoinIterator(
+   */
+  private def updateTrackingMask(rightGatherMap: LazySpillableGatherMap): Unit = {
+    // Get the current tracking table, or all true table to start with
+    val rightTrackingTable = rightTracker.map { spillableBatch =>
```
This leaks if something throws before we get to the `withResource(rightTrackingTable)` below.  Seems like this is being created too early.

jbrennan333 replied:
Fixed.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -511,6 +507,208 @@ class ConditionalHashJoinIterator(
+      trueColumnTable(numRightRows)
+    }
+    // Filter the right gather map to remove invalid indices
+    val numGatherMapRows = rightGatherMap.getRowCount.toInt
```
Are we guaranteed this doesn't truncate?  Unlike cudf columns, gather maps can be longer than 2^31 entries (which is why a gather map is not a column).  Unsure if we've already split the gather map to "fit" at this point yet or if this is the "raw" gather map that came out of a left outer join call in libcudf and has yet to be split if it's too large (e.g.: left outer join exploded).

jbrennan333 replied:
This is called via `createGatherer` from `SplittableJoinIterator.setupNextGatherersetupNextGatherer`.
It is called after the code that splits the stream side if we think the num rows will be too big.  But I was thinking this would be ok since this is the gather map for the build side, which should already be limited.  Is this an incorrect assumption?

jbrennan333 replied:
I didn't make a change for this.  Do you agree with my reasoning?   We are always using the build side gather map for this.

jlowe replied:
yes we're fine here.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -511,6 +507,208 @@ class ConditionalHashJoinIterator(
+    val updatedTrackingTable = {
+      withResource(rightTrackingTable) { trackingTable =>
+        withResource(Scalar.fromBool(false)) { falseScalar =>
+          withResource(filteredGatherMap) { filteredMap =>
```
filteredGatherMap can leak if `fromBool` just above throws

jbrennan333 replied:
I will fix this.

jbrennan333 replied:
Fixed.

revans2 commented on spark2-sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -133,13 +133,13 @@ object GpuHashJoin {
   /** Determine if this type of join supports using the right side of the join as the build side. */
   def canBuildRight(joinType: JoinType): Boolean = joinType match {
-    case _: InnerLike | LeftOuter | LeftSemi | LeftAnti | _: ExistenceJoin => true
+    case _: InnerLike | LeftOuter | LeftSemi | LeftAnti | FullOuter | _: ExistenceJoin => true
```
This is confusing. It was confusing before to have `FullOuter` have a left build. Should we have a follow on to pick the best side instead? Also could we add some docs about why this is now build right?

jbrennan333 replied:
In this initial version, I changed this to match LeftOuter join because I was implementing it as a Left Outer join, so I figured I should use the same constraints.  Based on these comments, I agree full join needs to be able to build left of right.  I'm trying to decide if that means I need to use a Right Outer join when building left and Left Outer join when building right?

jbrennan333 replied:
The code now allows build left or build right.

revans2 reviewed (commented):
I essentially had the same comments as @jlowe

jbrennan333 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala:
```diff
@@ -73,6 +73,8 @@ abstract class AbstractGpuJoinIterator(
    */
   protected def setupNextGatherer(): Option[JoinGatherer]
 
+  protected def hasFinalBatch(): Option[ColumnarBatch] = None
```
Good suggestion.  I will rename.

jbrennan333 replied:
This is changed.

jbrennan333 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -136,13 +136,13 @@ object GpuHashJoin extends Arm {
 
   /** Determine if this type of join supports using the left side of the join as the build side. */
   def canBuildLeft(joinType: JoinType): Boolean = joinType match {
-    case _: InnerLike | RightOuter | FullOuter => true
```
Agree (as per my comment above).  If building left, I'm thinking I'll need to use Right Outer join instead of Left Outer Join?

revans2 replied:
To be clear here, this is not going to make any difference until we actually fix how we pick the build side for a sort merge join translation.

```
  val buildSide: GpuBuildSide = if (GpuHashJoin.canBuildRight(join.joinType)) {
    GpuBuildRight
  } else if (GpuHashJoin.canBuildLeft(join.joinType)) {
    GpuBuildLeft
  } else {
    throw new IllegalStateException(s"Cannot build either side for ${join.joinType} join")
  }
```

Perhaps there are cases if we enable hash joins in Spark that it might show up, but unless we can force this to happen in testing we are going to have one code path that is not tested.

jbrennan333 replied:
The logic for `sortMergeJoin` favors `GpuBuildRight` without looking at the join side from spark.  But I think in most other cases we use `getGpuBuildSide` and then match on that to check if we can build on that side.  So I think I do need to handle the buildLeft case as well.

jbrennan333 replied:
I have updated to support build left/right, but I didn't change this sortMergeJoin buildside code.  Should I go ahead and do that as part of this PR, or file a followup?

jbrennan333 commented:
I have addressed some of the review comments and I merged up the branch.
Still todo:

1. Allow left or right side build.
2. Change coalesce children logic to allow multiple batches on stream side.

jbrennan333 commented:
I have updated this to allow for build left and build right.  All integration tests pass, and I've also run nds2.0 on my desktop - q51 and q97 both have full joins in them.  I've verified correctness for those.
I will also run nds2.0 on spark2a and do a benchmark comparison.

I have not yet updated any integration tests.  I will make a pass through to see if I think any need to be updated.  If you have suggestions for new or improved integration tests for this please let me know.  Or suggestions for additional testing.

jbrennan333 commented:
build

jbrennan333 commented:
Attaching some images from the event logs for q51/q97 before and after this change that shows how the full outer hash join changes.  In both cases the build side switches, and GpuCoalesceBatches are removed from the streaming side.

q51-orig-full-outer-join-buildleft
![q51-orig-full-outer-join-buildleft](https://user-images.githubusercontent.com/69316431/212128825-5c10d442-8ba1-4215-af39-dfe6986c21f5.png)

q51-bfhj-full-outer-join-buildright
![q51-bfhj-full-outer-join-buildright](https://user-images.githubusercontent.com/69316431/212128816-998ddf8c-1b38-46ba-a5f1-236547884147.png)

q97-orig-full-outer-join-buildleft
![q97-orig-full-outer-join-buildleft](https://user-images.githubusercontent.com/69316431/212128842-0ad2fd04-2835-435d-90cc-d76e331b1d10.png)

q97-bfhj-full-outer-join-buildright
![q97-bfhj-full-outer-join-buildright](https://user-images.githubusercontent.com/69316431/212128835-152a5ba4-f434-42a7-8ecc-246ef337d3e9.png)

jbrennan333 commented:
I ran a benchmark comparison (5 iterations) for nds2.0 on the spark2a cluster.  With the full nds power run it showed a small regression for query95, but no overall regression (or improvement).  I ran 10 iterations of the experiment for just q95 and it found no significant change, so I think the query95 result in the full run was spurious.  So for nds at 3TB, I am not seeing any significant performance impact from this change.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastHashJoinExec.scala:
```diff
@@ -132,13 +132,7 @@ case class GpuBroadcastHashJoinExec(
-    case (FullOuter, GpuBuildRight) => Seq(RequireSingleBatch, null)
-    case (_, _) => Seq(null, null)
-  }
+  override def childrenCoalesceGoal: Seq[CoalesceGoal] = Seq(null, null)
 
```
Nit: This is already the default.
```suggestion
```

jbrennan333 replied:
I will remove this.

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -511,6 +510,235 @@ class ConditionalHashJoinIterator(
+    }
+    // Update all hits in the gather map with false (no longer needed) in the tracking table
+    val updatedTrackingTable = {
+      withResource(filteredGatherMap) { filteredMap =>
```
We leak filteredGatherMap if something above throws (e.g.: spillableBatch.releaseBatch or trueColumnTable)

jbrennan333 replied:
Good catch.  Thanks!

jlowe commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:
```diff
@@ -511,6 +510,235 @@ class ConditionalHashJoinIterator(
+      trueColumnTable(numBuiltRows)
+    }
+    // Update all hits in the gather map with false (no longer needed) in the tracking table
+    val updatedTrackingTable = {
+      withResource(filteredGatherMap) { filteredMap =>
```
Nit: Can save on some unnecessary indentation here by just assigning from the result of `withResource` directly, e.g.:
```scala
    val updatedTrackingTable = withResource(filteredGatherMap) { filteredMap =>
```
Also applies to where an extra block was used below for `builtSideTracker`

jbrennan333 replied:
Good suggestion.

jlowe commented:
Does this PR still need to be a draft?  Tests and benchmark results indicate it's ready to go once it passes review.

jbrennan333 commented:
> Does this PR still need to be a draft? Tests and benchmark results indicate it's ready to go once it passes review.
Thanks for the reviews @jlowe!  I was waiting until I got the full implementation up and had an initial review to make sure the approach for handling left/right build sides was correct.   I will move it out of draft.

jbrennan333 commented:
build

jbrennan333 commented:
build

jbrennan333 commented:
The databricks and 340 failures appear to be unrelated.  Should I rerun tests with [databricks] removed?

jbrennan333 commented:
This is the failure in the databricks part of the build ci:
```
java.lang.NullPointerException
at org.apache.spark.storage.BlockManagerMasterEndpoint.org$apache$spark$storage$BlockManagerMasterEndpoint$$register(BlockManagerMasterEndpoint.scala:553)
at org.apache.spark.storage.BlockManagerMasterEndpoint$$anonfun$receiveAndReply$1.applyOrElse(BlockManagerMasterEndpoint.scala:117)
        at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:103)
        at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213)
	at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100)
	at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75)
	at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable.$anonfun$run$1(SparkThreadLocalForwardingThreadPoolExecutor.scala:104)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
 	at org.apache.spark.util.threads.SparkThreadLocalCapturingHelper.runWithCaptured(SparkThreadLocalForwardingThreadPoolExecutor.scala:68)
	at org.apache.spark.util.threads.SparkThreadLocalCapturingHelper.runWithCaptured$(SparkThreadLocalForwardingThreadPoolExecutor.scala:54)

```

jbrennan333 commented:
build

jbrennan333 commented:
Thanks for the reviews @jlowe and @revans2!
